packages/formspree-react/src/useForm.ts

```diff
@@ -5,50 +5,51 @@ import { useElements, CardElement, useStripe } from '@stripe/react-stripe-js';
 import { useFormspree } from './context';
 import type { ExtraData } from './types';
 import { version } from '../package.json';
-import type {
-  Client,
-  SubmissionResponse,
-  SubmissionData,
-  ErrorBody,
-  FormError,
+import {
+  appendExtraData,
+  type Client,
+  type FieldValues,
+  type SubmissionData,
+  type SubmissionSuccessResult,
+  type SubmissionErrorResult,
 } from '@formspree/core';
 
 type FormEvent = React.FormEvent<HTMLFormElement>;
 
-type SubmitHandler = (
-  submissionData: FormEvent | SubmissionData
-) => Promise<SubmissionResponse>;
+type SubmitHandler<T extends FieldValues> = (
+  submissionData: FormEvent | SubmissionData<T>
+) => Promise<void>;
 
 type ResetFunction = () => void;
 
-export type TUseForm = [
+export type TUseForm<T extends FieldValues> = [
   {
-    result: SubmissionResponse | null;
+    errors: SubmissionErrorResult<T> | null;
+    result: SubmissionSuccessResult | null;
     submitting: boolean;
     succeeded: boolean;
-    errors: FormError[];
   },
-  SubmitHandler,
+  SubmitHandler<T>,
   ResetFunction
 ];
 
 const isEvent = (data: FormEvent | SubmissionData): data is FormEvent => {
   return (data as FormEvent).preventDefault !== undefined;
 };
 
-const useForm = (
+const useForm = <T extends FieldValues>(
```

Comment:
For this one, user needs to provide the type argument, if they want to have the type information available in the `handleSubmit` function as well as the `SubmissionErrorResult`.

If the type argument `T` is not provided, the default `FieldValues` (`Record<string, string | ...>`) would be used.

Maybe there's a typescript pattern that can make this type inferred (maybe using `infer` keyword) but I don't know how. 😅

Reply:
So how is the typical typsecript user expected to invoke useForm? With or without template variable?

Reply:
Can you give an example code snippet of using useForm with an explicit type parameter?

Reply:
## An example of using it with `FormEvent`.

```tsx
export function MyForm() {
  const [{ errors, submitting }, handleSubmit] = useForm('formkey-1234');
  return (
    <form onSubmit={handleSubmit}>
      <input type="email" name="email" />
      <ValidationError errors={errors} field="email" />
      <button disabled={submitting}>
        {submitting ? 'Submitting…' : 'Submit'}
      </button>
    </form>
  );
}
```

If we don't provide the type argument, it'll use `FieldValue` which is `Record<string, ...>`.

<img width="769" alt="image" src="https://github.com/formspree/formspree-js/assets/3942264/54c56e10-44c4-4cdf-99b6-3244d55a5562">

## An example using it with a POJO `FormData`

```tsx
export function MyForm() {
  const [email, setEmail] = useState('');
  const [{ errors, submitting }, handleSubmit] = useForm<{ email: string }>(
    'formkey-1234'
  );
  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();
        handleSubmit({ email });
      }}
    >
      <input
        name="email"
        value={email}
        onChange={(e) => setEmail(e.target.value)}
      />
      <ValidationError errors={errors} field="email" />
      <button disabled={submitting}>
        {submitting ? 'Submitting…' : 'Submit'}
      </button>
    </form>
  );
}
```

In this case, since the type is passed, the error fields are strongly typed.

<img width="804" alt="image" src="https://github.com/formspree/formspree-js/assets/3942264/26781a67-d4f5-44ca-9163-cf68ce13b4c4">

I hope there's a better way to infer the type from the calling position of the `handleSubmit` function but like previously mentioned, I couldn't figure it out (not sure it's possible in typescript).

Reply:
What do you mean by "With or without template variable?"?